8356173: Remove ThreadCritical

coleenp opened the pull request:
Updated the description in the bug.  This removes the last use of ThreadCritical and replaces it with a global PlatformMutex lock.
Tested with tier1-4, and tier1 on all Oracle-supported OSs.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8356173](https://bugs.openjdk.org/browse/JDK-8356173): Remove ThreadCritical (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25072/head:pull/25072` \
`$ git checkout pull/25072`

Update a local copy of the PR: \
`$ git checkout pull/25072` \
`$ git pull https://git.openjdk.org/jdk.git pull/25072/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25072`

View PR using the GUI difftool: \
`$ git pr show -t 25072`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25072.diff">https://git.openjdk.org/jdk/pull/25072.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25072#issuecomment-2859913360)
</details>

dholmes-ora commented on src/hotspot/share/nmt/nmtUsage.cpp:
```diff
@@ -52,9 +52,9 @@ void NMTUsage::walk_thread_stacks() {
 
 void NMTUsage::update_malloc_usage() {
-  // Thread critical needed keep values in sync, total area size
+  // Lock needed keep values in sync, total area size
```
```suggestion
  // Lock needed to keep values in sync, total area size
```

dholmes-ora commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -459,6 +459,9 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
   os::init();
 
+  // Initialize memory pools
+  Arena::initialize_chunk_pool();
```
What code first uses a `ChunkPool` or the `ChunkPoolLocker`? (It isn't obvious at what point the NMT code may execute during early initialization.)

coleenp replied:
Yes, this is needed early.  We create the ResourceArea arena when creating a JavaThread before the JavaThread is completely created so Thread::current() doesn't exist in this case.  There are also other early initializations.

There's two things that make us unable to have a static PlatformMutex.  1. on macosx, the PlatformMutex is implemented using an indirection the macosx dll load crashes trying to initialize it.

```
#  assert(status == 0) failed: error EINVAL(22), freelist lock

V  [libjvm.dylib+0xea68c4]  PlatformMutex::PlatformMutex()+0xfc
V  [libjvm.dylib+0x269c34]  _GLOBAL__sub_I_arena.cpp+0x38
C  [dyld+0x22a24]  invocation function for block in dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const::$_0::operator()() const+0xa8
C  [dyld+0x680f4]  invocation function for block in dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const+0xac
C  [dyld+0x5b668]  invocation function for block in dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const+0x1f0
C  [dyld+0x22fc]  dyld3::MachOFile::forEachLoadCommand(Diagnostics&, void (load_command const*, bool&) block_pointer) const+0x12c
C  [dyld+0x5a6a0]  dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const+0xc0
C  [dyld+0x5d188]  dyld3::MachOFile::forEachInitializerPointerSection(Diagnostics&, void (unsigned int, unsigned int, bool&) block_pointer) const+0xa0
C  [dyld+0x67de8]  dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) 
```

Even if I disable the indirect PlatformMutex workaround for the macosx bug JDK-8218975, the underlying pthread_mutex is not initialized before this code uses it.

```
#  assert(status == 0) failed: error EINVAL(22), mutex_init
```

So I don't think PlatformMutex can be a static variable, which is why I initialize it early in vm startup.

coleenp replied:
I added the initialization of the mutex after os::init().  Assuming that pthread_mutex will be initialized at that point.

tstuefe replied:
@coleenp But surely one can statically allocate a pthread_mutex_t and assign PTHREAD_MUTEX_INITIALIZER, no? I mean that's Unix 101. MacOS should be able to do that at least.

Looking at PlatformMutex, maybe it is not the right tool for the job. One needs a simple stupid mutex class like this:

```
struct SimpleMutex {
    pthread_mutex_t _m;
    SimpleMutex() : _m(PTHREAD_MUTEX_INITIALIZER) {}
    lock() -> do pthread_mutex_lock
    unlock() -> do pthread_mutex_unlock
};
```

etc.

tstuefe replied:
Note that such a simple solution would also reduce the number of indirections when using this mutex. One less pointer load, one less memory access.

coleenp replied:
PlatformMutex is the simple locking scheme that we use throughout the VM, so it is the right tool for the job. It is actually the same as your SimpleMutex only it had code that deals with the old macosx bug (that should be removed not with this RFE).  Actually ThreadCritical is your SimpleMutex except has to be reimplemented on both windows and posix, which again is already the same as PlatformMutex.

>  I mean that's Unix 101. MacOS should be able to do that at least.

My crash suggests otherwise.  Maybe it's a bug in PlatformMutex.  I'd rather fix that.

kimbarrett replied:
https://bugs.openjdk.org/browse/JDK-8220805 is still lingering? Looks like it was decided to fix it some time
ago, but that still hasn't happened.  That's likely what is causing the crash @coleenp encountered.

kimbarrett replied:
@tstuefe We can't rely on PlatformMutex being initialized that way, because the Windows implementation
requires a _call_ to `InitializeCriticalSection`.

coleenp replied:
I didn't know this about Windows.  The Windows code seemed fine with the static PlatformMutex.  The posix code calls pthread_mutex_init with an parameter that's initialized in os::init() -> pthread_init_common();

If I call pthread_init_common() in all the places that  _mutexAttr and condAttr are used, the macosx platform is happier.  And I can make PlatformMutex static and remove that initialize_chunk_pool() function.  I don't think this is a nice change though.

kimbarrett replied:
For Windows, a static PlatformMutex introduces a global variable with a non-trivial constructor,
which we avoid because of ye old "static initialization order fiasco".  I don't think we should
make this PlatformMutex statically allocated.  We should retain initialize_chunk_pool().
(Especially because the lock-free chunk-pool idea that @jdksjolen  and I have discussed likely
wants an initialization function.)

coleenp replied:
Good.  initialize_chunk_pool() is at the right place after os::init which for posix anyways initializes global variables that pthread_mutex_init needs.

dholmes-ora reviewed (commented):
This seems reasonable but I have a few minor queries.

It was surprising how many files included `threadCritical.hpp` for no apparent reason.

Thanks

tstuefe commented on src/hotspot/share/nmt/nmtUsage.cpp:
```diff
@@ -52,9 +52,9 @@ void NMTUsage::walk_thread_stacks() {
   // is deducted from mtChunk in the end to give correct values.
-  ThreadCritical tc;
+  ChunkPoolLocker lock;
   const MallocMemorySnapshot* ms = MallocMemorySummary::as_snapshot();
```
I'd scope this to the as_snapshot call, the subsequent code does not need the locking (lesser chance of accidental recursive locking)

coleenp replied:
done.

tstuefe commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -31,12 +31,30 @@
+// One global static mutex for chunk pools.
+// It is used very early in the vm initialization, in allocation
+// code and other areas.  For many calls, the current thread has not
+// been created so we cannot use Mutex.
+static PlatformMutex* _global_chunk_pool_mutex = nullptr;
+
+void Arena::initialize_chunk_pool() {
+  _global_chunk_pool_mutex = new PlatformMutex();
+}
```
Why do we even need dynamic initialization here? I thought that is the tradeoff with PlatformMutex: you forego deadlock checks etc, but you gain safety wrt initialisation.
That would also save some instructions.

jdksjolen replied:
You can use `Deferred<T>`, but as `PlatformMutex` is at "the bottom" of the initialization order it's not necessary, as Thomas notes. I'd be happy to see it use `Deferred<T>`, as it can help with debugging if PlatformMutex ever changes.

coleenp replied:
Not really crazy about the Deferred<> name and that change is still under consideration.

coleenp replied:
Deferred<> doesn't help me because it still has to call the initialize function, which is the thing we don't want to have to call.

kimbarrett replied:
As I said in another comment, Windows appears to require a call to an initialization function.  So this seems
like a good candidate for Deferred<> to me, removing some relatively boiler-plate code.

coleenp replied:
I'm not really sold on what the purpose of Deferred is tbh.  Seems like it would add some more boilerplate since I the initialization function is only one obvious line.

kimbarrett replied:
The initialization lines are similar one-liners.
Deferred<> access already had the initialized check built-in, so don't need the explicit assert.
Deferred<> avoids the pointer indirection.
Deferred<> avoids dynamic allocation.

coleenp replied:
I still fail to see the point of this.  dynamic allocation isn't something that's a problem for this object.  It may be for the NMT types that use this wrapper but not for this object.  You cannot write a microbenchmark to show the effect of this pointer indirection in this code path.  If you could, there's something seriously wrong with our algorithms which should be fixed instead.  The initialization check is an obvious one liner, not executable in production.

The DeferredWhat? wrapper is extra typing, that requires you to go to some other place to figure out that among all the casting, it's doing a placement new.  For no reason other than to give you more code to navigate to.

kimbarrett replied:
Clearly we disagree about the utility of Deferred. I won't insist on it being
used here, even though I think it's a fine use-case. I think this comment from
the PR best describes the motivation:
https://github.com/openjdk/jdk/pull/24689#issuecomment-2823556707
Unfortunately, that comment is kind of buried down in the middle of the PR.

But some of the criticisms seem incorrect. For example, using Deferred is less
typing, even without including the assert against multiple initialization that
is missing from the current code. (That's one of the problems with
boiler-plate - it's easy to forget or mess up some of it.) And I don't know
what you mean by "all the casting", as there is precisely one (const) cast in
Deferred. Maybe you are remembering earlier versions from the PR that added
it? (Some versions did have lots of casts.)

coleenp replied:
As an aside, I like the StableValue name better than Deferred, as the latter means nothing to me.  So changing to Deferred would just defer my attention to another file to find out what Deferred means, when this is literally 5 lines of code in front of me.  So I would not like to change it.

tstuefe reviewed (commented):
I like that we use PlatformMutex, not Mutex; I think that is a very reasonable approach. We forego deadlock detection but gain independence wrt initialization order.

Note that if you use PlatformMutex, there is no reason to dynamically allocate the mutex; just allocate as a global variable.

There is a small danger associated with this proposal in that ThreadCritical was allowing for recursive entry, PlatformMutex does not. On Posix it uses pthread_mutex_xxx, which on some platforms has the ability to work recursively; but on Windows we use critical sections, there its not possible. (we could use Windows Mutex instead).

Why not add a debug-only information to the Locker class to hold the owning thread id, and assert on enter with my own?

But actually the usage of the Locker is rather safe; The only real concern with potential recursive entry is with guarding os::free, and with potential deadlocks should we crash inside a signal handler (error handling or AsyncGetCallTrace or that new "CPU Time Profiling for JFR" JEP). 

For these signal handler usages, I had this idea: https://bugs.openjdk.org/browse/JDK-8349578 - a double-buffering approach where we, upon entering the signal handler, open up a new ResourceArea for the current thread, and return to the original RA upon leaving signal handling.

About guarding os::free, that is needed for a dreary issue (see https://bugs.openjdk.org/browse/JDK-8325890). In short, it is needed to get stable arena memory readings. 

However, I would love to just get rid of arena memory accounting altogether; the solution is very hackish and probably not needed anymore. The only heavy user of chunks causing repeated problems that caused me to use NMT at customers was the JIT. And I added the compilation memory statistic last year, so we have not a much much better tool to investigate JIT arena usage.

We also have an issue https://bugs.openjdk.org/browse/JDK-8333151, which investigates whether we can get rid of chunkpool altogether; see the ratio given in that issue description. This gives some possible performance benefits apart from simplification, but comes at a significant risk of backward compatibility problems since we are then much more subject to memory retention policy of the glibc.

---

Bottom line, I like this PR and if possible would add some debug-only way to catch recursive entry.

coleenp commented on src/hotspot/share/nmt/mallocTracker.cpp:
```diff
@@ -61,10 +62,10 @@ void MemoryCounter::update_peak(size_t size, size_t cnt) {
   // copy is going on, because their size is adjusted using this
   // buffer in make_adjustment().
-  ThreadCritical tc;
+  ChunkPoolLocker lock;
```
Yes, ThreadCritical preceded the PlatformMutex locks, which were added later.  We had ThreadCritical (iirc) for a limitation in the Windows malloc code, which is why it was around all the os::malloc calls once. NMT just built upon that since it was there.
We cannot use regular Mutex because Mutex requires Thread::current() which doesn't exist at this point. The NMT code might be able to use a lock around the code that deletes arenas, as the comments suggest that it is trying to protect.  This could be a regular Mutex because the arena code is deleted individually and purged well after startup time.  I left this as a shared lock, but it could be further improved to use a nmt-delete-lock Mutex.

coleenp commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -198,7 +216,8 @@ void ChunkPool::deallocate_chunk(Chunk* c) {
     pool->return_to_pool(c);
   } else {
-    ThreadCritical tc;  // Free chunks under TC lock so that NMT adjustment is stable.
+    // Free chunks under NMT lock so that NMT adjustment is stable.
```
This could use an NMT lock instead (had a version with this once).  I fixed the comment to reflect this change.

coleenp commented:
> For these signal handler usages, I had this idea: https://bugs.openjdk.org/browse/JDK-8349578 - a double-buffering approach where we, upon entering the signal handler, open up a new ResourceArea for the current thread, and return to the original RA upon leaving signal handling.
>
>About guarding os::free, that is needed for a dreary issue (see https://bugs.openjdk.org/browse/JDK-8325890). In short, it is needed to get stable arena memory readings.

I don't think this patch makes either of these NMT problems worse.  The ChunkPoolLocker (except for the call to os::free()) holds the lock for a very short interval.  Getting interrupted here is the same as getting interrupted in any critical section in the vm.

The os::free() calls could be wrapped in an NMT specific lock.  Or the mtChunk accounting could just use atomic operations - they don't already?

I realize with this change I'm interfering with some ongoing NMT work.  I apologize for that. I was trying to get rid of a case of a raw pthread_mutex call that we were looking at in some other context, and to remove the last remnants of the once pervasive ThreadCritical lock.

kimbarrett commented on src/hotspot/share/nmt/nmtUsage.cpp:
```diff
@@ -52,10 +52,13 @@ void NMTUsage::walk_thread_stacks() {
+  {
+    ChunkPoolLocker lock;
+    ms = MallocMemorySummary::as_snapshot();
+  }
```
Shortening the scope of the lock seems contrary to the comment explaining its presence.

kimbarrett replied:
I missed the discussion that suggested this shortening before commenting.  I guess the snapshot
is what provides the atomicity for the remainder, so shortening the scope seems okay after all.

kimbarrett commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -31,12 +31,34 @@
+// code and other areas.  For many calls, the current thread has not
+// been created so we cannot use Mutex.
+static PlatformMutex* _global_chunk_pool_mutex = nullptr;
+static volatile int _recursion_count = 0;
```
Why is this volatile?  It's only accessed in the context of the lock.  Also, I think it probably doesn't
help for POSIX, since we're using `PTHREAD_MUTEX_NORMAL`, for which attempting recursion
"shall deadlock", with no error check.  (If you want error checking, you need to use
`PTHREAD_MUTEX_ERRORCHECK` and check the result of the lock attempt.)

I think the Windows critical sections that are used to implement PlatformMutex do allow recursion,
and maybe we should have this kind of guard to prevent recursive use there, to be consistent with
the restriction imposed by the POSIX implementation.  (Or change the POSIX implementation to
use `PTHREAD_MUTEX_RECURSIVE`, but I don't think we should do that.)

Maybe the POSIX PlatformMutex should use `PTHREAD_MUTEX_ERRORCHECK` in debug
builds.  But that's a separate issue.

coleenp replied:
You're right, it doesn't need to be volatile. I want the recursion checking to be platform independent for this lock, and not affect the other PlatformMutex/Monitors which are outside the scope of this change and hopefully are what we want already.

kimbarrett replied:
But the recursion checking isn't platform independent.  For POSIX it does nothing useful and is just useless
clutter providing a false sense of security.  I think it should be removed, and consider (as a separate thing)
adding a recursion assert to the Windows implementation.

And it should be debug-only there; there's no need to change the counter in a release build, since the
check is an assert.  It also doesn't need to be a counter; it could be a bool flag.  But that's really about a
future Windows-check.  It just shouldn't be here.

dholmes-ora replied:
I have to agree with Kim here, you can't implement a recursion check this way using a count, which has to be protected by the mutex itself. You would need to maintain an owner field, using the platform specific thread id type, which is checked before trying to acquire the mutex.

coleenp replied:
oh yeah now I see it. That's not going to work unless PlatformMutex is already recursive.

kimbarrett commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -31,12 +31,30 @@
+// It is used very early in the vm initialization, in allocation
+// code and other areas.  For many calls, the current thread has not
+// been created so we cannot use Mutex.
+static PlatformMutex* _global_chunk_pool_mutex = nullptr;
```
This isn't a class data member, so a leading underscore shouldn't be used.
https://github.com/openjdk/jdk/blame/b6b5ac1ef9042ed62a8358aa6943b8dc87dcf0ab/doc/hotspot-style.md#L260-L262

Maybe this falls under:
https://github.com/openjdk/jdk/blame/b6b5ac1ef9042ed62a8358aa6943b8dc87dcf0ab/doc/hotspot-style.md#L223-L224
though I think that's rare usage for file-scoped static variables.  Just
dropping the leading underscore would probably be the most consistent with
other code.

coleenp replied:
Hm TIL - I've always used underscores for global static variables.

tstuefe commented:
@kimbarrett @coleenp I proposed to use simple global initialization because it makes the code independent from initialization order worries in this sort of very low level very early invoked code. That was the main benefit I had in mind. It is not an immediate concern (we have NMT preinit to deal with early VM time) but every avoided dependency is good.

The problem with global initialization was on POSIX that pthread_mutex_init() crashes when used at lib loading time on MacOs. The fix for this is to use a static PTHREAD_MUTEX_... initializer, which is also simpler.

On Windows, one initializes CriticalSections with InitializeCriticalSection(), but that can happen at initialization time in a constructor of a global object. I have done this combination (InitializeCriticalSection on windows, pthread_mutex_t with static initializer on Posix) a lot in many projects, it just works and is simple.

I am fine with whatever Coleen decides. I don't want to bikeshed this to death, just explain my reasoning.

P.S. Windows critical sections don't offer recursion checks.

kimbarrett commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -40,19 +40,19 @@
 // code and other areas.  For many calls, the current thread has not
 // been created so we cannot use Mutex.
-static PlatformMutex* _global_chunk_pool_mutex = nullptr;
+static PlatformMutex* GlobalChunkPoolMutex = nullptr;
```
That naming style also seems consistent with some of the code base and with the style guide.  So okay.

kimbarrett reviewed (approved):
Looks good.

coleenp commented:
Thanks Kim.

dholmes-ora reviewed (approved):
I think the dust has settled on this now. :)

LGTM.

Thanks

jdksjolen reviewed (approved):
This seems fine by me as well. Thanks for removing some more code :-).

coleenp commented:
@tstuefe To make the global initialization work, at least on posix, I had to fix the os_posix.cpp code because it was using an uninitialized variable to call pthread_mutex_init.  If I fixed that (check an initialized flag and use the default if it hadn't been initialized), I still had to call pthread_init_common() in os_posix.cpp to guard all the places the code uses these values for _mutexAttr and _condAttr. This did work and enabled me to make the GlobalChunkPoolMutex static.  But then I'd have to fix the windows code, and I don't know windows.  So calling an initialization function when I know the os code is initialized is the platform independent way that I implemented this.  It may be that there might be other reasons to fix this OS code to allow static initialization, but that wasn't the scope of this change.  I just wanted ThreadCritical to be removed.  We were looking at this in the context of other locking changes that are forthcoming, and we want os level locking to be more uniform under PlatformMutex.

Thanks for reviewing and you comments Kim, David, Johan and Thomas.
/integrate

tstuefe commented:
> It may be that there might be other reasons to fix this OS code to allow static initialization, but that wasn't the scope of this change. I just wanted ThreadCritical to be removed. We were looking at this in the context of other locking changes that are forthcoming, and we want os level locking to be more uniform under PlatformMutex.

Certainly, sorry for delaying this simple fix. We can always improve upon it later if we want.
